src/array_api_extra/_funcs.py

```diff
@@ -548,3 +559,279 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+        """Allow for the alternate syntax ``at(x)[start:stop:step]``,
+        which looks prettier than ``at(x, slice(start, stop, step))``
+        and feels more intuitive coming from the JAX documentation.
+        """
+        if self.idx is not _undef:
+            msg = "Index has already been set"
+            raise ValueError(msg)
+        self.idx = idx
+        return self
+
+    def _common(
+        self,
+        at_op: str,
+        y: Array = _undef,
+        /,
+        copy: bool | None = True,
+        xp: ModuleType | None = None,
+        _is_update: bool = True,
+        **kwargs: Untyped,
+    ) -> tuple[Untyped, None] | tuple[None, Array]:
+        """Perform common prepocessing.
+
+        Returns
+        -------
+        If the operation can be resolved by at[], (return value, None)
+        Otherwise, (None, preprocessed x)
+        """
+        if self.idx is _undef:
+            msg = (
+                "Index has not been set.\n"
+                "Usage: either\n"
+                "    at(x, idx).set(value)\n"
+                "or\n"
+                "    at(x)[idx].set(value)\n"
+                "(same for all other methods)."
+            )
+            raise TypeError(msg)
+
+        x = self.x
+
+        if copy is True:
+            writeable = None
+        elif copy is False:
+            writeable = is_writeable_array(x)
+            if not writeable:
+                msg = "Cannot modify parameter in place"
+                raise ValueError(msg)
+        elif copy is None:  # type: ignore[redundant-expr]
+            writeable = is_writeable_array(x)
+            copy = _is_update and not writeable
+        else:
+            msg = f"Invalid value for copy: {copy!r}"  # type: ignore[unreachable]  # pyright: ignore[reportUnreachable]
+            raise ValueError(msg)
+
+        if copy:
+            try:
+                at_ = x.at
+            except AttributeError:
+                # Emulate at[] behaviour for non-JAX arrays
+                # with a copy followed by an update
+                if xp is None:
+                    xp = array_namespace(x)
+                # Create writeable copy of read-only numpy array
+                x = xp.asarray(x, copy=True)
+                if writeable is False:
+                    # A copy of a read-only numpy array is writeable
+                    writeable = None
+            else:
+                # Use JAX's at[] or other library that with the same duck-type API
+                args = (y,) if y is not _undef else ()
+                return getattr(at_[self.idx], at_op)(*args, **kwargs), None
+
+        if _is_update:
+            if writeable is None:
+                writeable = is_writeable_array(x)
+            if not writeable:
+                # sparse crashes here
+                msg = f"Array {x} has no `at` method and is read-only"
+                raise ValueError(msg)
+
+        return None, x
+
+    def get(self, **kwargs: Untyped) -> Untyped:
```

Comment:
```suggestion
    def get(self, /, **kwargs: Untyped) -> Untyped:
```

Reply:
given the `kwargs.get("copy")`, you at least know that `*, copy: bool | None = None`, right?

Reply:
Yes but I wanted the signature to be consistent with the other methods in the sphinx doc.
I'll probably change all methods to have an explicit `copy` and `xp` parameters.

Reply:
> I'll probably change all methods to have an explicit `copy` and `xp` parameters.

Once you do, it could help avoid duplication by using a `TypedDict` and `Unpack` (which requires `typing_extensions` on `python < 3.11`) in that case.